### PR TITLE
feat: add FFI function for get_earliest_delta

### PR DESCRIPTION
**Pull request by thinh2**
## What changes are proposed in this pull request?

This PR added an FFI implementation of the history_manager::get_earliest_delta function, allowing connectors written in other languages (C/C++/Java/...) querying the earliest commit version in the delta_log of the Delta table. Connectors can use the FfiHistoryCommitType to control the returned history commit type.

- FfiHistoryCommitType::Published will find the earliest commit available in the delta_log. However, it does not guarantee the table can be constructed from that version.
- FfiHistoryCommitType::Recreatable will find the earliest commit version that the table can be constructed from and replay forward to the lastest version.

## How was this change tested?

- New unit test was added, verifying the new FFI function return the expected result.

### CODE REVIEW

File: ffi/src/lib.rs [hunk 1]
```diff
@@ -1514,6 +1591,132 @@ mod tests {
         Ok(())
     }
 
+    enum EarliestCommitTableSetupScenario {
+        /// Filesystem-only table with a single v0 metadata commit.
+        Filesystem,
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: nit: we should make these scenarios more self describing.

FilesystemV0Commit
CatalogManagedV0Commit
FilesystemV4CheckpointWithEarlierCommit
NoCommits

etc.

For the empty case, is it a file system or catalog managed commit? can we use it to also test out the CM v0 not published error case?

File: ffi/src/error.rs [hunk 2]
```diff
@@ -343,7 +345,8 @@ impl From<EngineExecError> for Error {
             | KernelError::ParseIntervalError
             | KernelError::ChangeDataFeedUnsupported
             | KernelError::ChangeDataFeedIncompatibleSchema
-            | KernelError::LiteralExpressionTransformError) => {
+            | KernelError::LiteralExpressionTransformError
+            | KernelError::LogHistoryError) => {
```
Comment (chiinlquah) [inline ffi/src/error.rs:349]: we have defined error for log history no? We should map it to LogHistoryError above instead.

**Review by chiinlquah — COMMENTED** [review]
Thanks for the change, some comments.

File: ffi/src/error.rs (hunk 2 above)
Comment (thinh2) [reply ffi/src/error.rs:349]: the `Error::LogHistory` is not a simple string. It is an Enum wrapper for multiple Error type inside, e.g `InvalidTimestampRange`, `TimestampOutOfRange`, ... . Converting the error string back to an enum required a complexed string parsing to find the correct internal error Enum type. That why I mapped it to the GenericError to preserve the error code and message first.

File: ffi/src/lib.rs (hunk 1 above)
Comment (thinh2) [reply ffi/src/lib.rs]: Updated the enum to self-describing scenario and remove the enum's doc comment. I also add the test for the catalog managed table with v0 not published using the empty delta_log setup.

File: ffi/src/lib.rs [hunk 3]
```diff
@@ -999,6 +1011,71 @@ pub unsafe extern "C" fn snapshot_timestamp(
...
 }
 
+/// Selects which commit type to return for the history_manager query. FFI-safe mirror of
+/// [`HistoryCommitType`].
+#[repr(C)]
+pub enum FfiHistoryCommitType {
+    /// Maps to [`HistoryCommitType::Published`].
+    Published,
+    /// Maps to [`HistoryCommitType::Recreatable`].
+    Recreatable,
```
Comment (nicklan) [inline ffi/src/lib.rs]: Let's be sure these map to a known value:

```suggestion
    Published = 0,
    /// Maps to [`HistoryCommitType::Recreatable`].
    Recreatable = 1,
```

**Review by nicklan — APPROVED** [review]
lgtm with one small comment. thanks

**Review by chiinlquah — APPROVED** [review]
